arch: arm: cortex_m: Update APIs to save and restore FPU context

msmttchr opened the pull request:
Update API to save and restore FPU context to address suspend to ram use case.

The API is updated as follow:
- Existing APIs `z_arm_save_fp_context` and `z_arm_restore_fp_context` have been changed to always save/restore FPU context regardless of `FPU_SHARING` setting
- The only use case of the original API has been updated to retain original behaviour 

I also noticed by reading code and ARM documentation that the FPU context is preserved across threads by hardware on Cortex-M devices (via the FPCCR.ASPEN register field).

See also #90001, https://github.com/zephyrproject-rtos/zephyr/pull/92847#discussion_r2195238617

msmttchr committed:
arch: arm: coterx_m: Added some #include in fpu.h

Added missing #include for used types uint32_t and bool
in file fpu.h

Signed-off-by: Michele Sardo <msmttchr@gmail.com>

wearyzen commented on include/zephyr/arch/arm/cortex_m/fpu.h:
```diff
@@ -19,5 +19,7 @@ struct fpu_ctx_full {
 
 void z_arm_save_fp_context(struct fpu_ctx_full *buffer);
 void z_arm_restore_fp_context(const struct fpu_ctx_full *buffer);
+void z_arm_save_shared_fp_context(struct fpu_ctx_full *buffer)
```
I don't see a reason to have this api if all its going to do is call z_arm_save_fp_context

msmttchr replied:
Ok, I have removed the new API and kept original behavior in `modules/trusted-firmware-m/interface/interface.c` file. Do you agree ?

wearyzen reviewed (changes requested):
Thank you for the change, could you please have a look at https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines and fix the commit message?

msmttchr commented on include/zephyr/arch/arm/cortex_m/fpu.h:
```diff
@@ -7,6 +7,9 @@
 #define ZEPHYR_INCLUDE_ARCH_ARM_CORTEX_M_FPU_H_
 
+#include <stdint.h>
+#include <stdbool.h>
```
> Why?

The header file uses uint32_t and bool types

stephanosio replied:
> I guess header files should include what they need and not rely on possible other includes done by user. Otherwise you will have errors if you change the include order.

That is correct.

msmttchr committed:
arch: arm: cortex_m: Modifed FPU save and restore helpers

z_arm_save_fp_context and z_arm_restore_fp_context
save and restore fpu context regardless of the
CONFIG_FPU_SHARING setting.

This modification is required to support suspend to ram use cases
where save and restore of FPU state is needed to ensure proper bahaviour
after wakeup.

Signed-off-by: Michele Sardo <michele.sardo@st.com>

msmttchr committed:
modules: trusted-firmware-m: Update in interface implementation

Following changes to arch/arm/core/cortex_m/fpu.c,
the dependency on CONFIG_FPU_SHARING is moved into this file.

Signed-off-by: Michele Sardo <michele.sardo@st.com>

msmttchr commented:
> Thank you for the change, could you please have a look at https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines and fix the commit message?

Done.

stephanosio commented on arch/arm/core/cortex_m/fpu.c:
```diff
@@ -16,7 +16,7 @@
 void z_arm_save_fp_context(struct fpu_ctx_full *buffer)
 {
-#if defined(CONFIG_FPU_SHARING)
+#if defined(CONFIG_FPU)
```
From the commit message:

> z_arm_save_fp_context and z_arm_restore_fp_context
save and restore fpu context regardless of the
CONFIG_FPU_SHARING setting.

Please add the rationale for this in the commit message.

msmttchr replied:
> From the commit message:
> 
> > z_arm_save_fp_context and z_arm_restore_fp_context
> > save and restore fpu context regardless of the
> > CONFIG_FPU_SHARING setting.
> 
> Please add the rationale for this in the commit message.

Done.
